File: backend/models/world.py
from datetime import datetime
from typing import Optional, Dict, Any
from sqlmodel import SQLModel, Field, JSON

class WorldLore(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: str = Field(index=True)
    project_id: Optional[int] = Field(default=None, index=True)
    
    # Modular Lore Data
    architecture: Optional[str] = Field(default=None)
    atlas: Optional[str] = Field(default=None)
    history: Optional[str] = Field(default=None)
    systems: Optional[str] = Field(default=None)
    culture: Optional[str] = Field(default=None)
    
    # Metadata
    full_lore_blob: Optional[str] = Field(default=None)
    metadata: Dict[str, Any] = Field(default_factory=dict, sa_type=JSON)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

File: src/services/api/world.ts
import { apiRequest } from '@/lib/api-utils';

const API_BASE = '/api/world';

export interface WorldLore {
  id?: number;
  user_id: string;
  project_id?: number;
  architecture: string | null;
  atlas: string | null;
  history: string | null;
  systems: string | null;
  culture: string | null;
  full_lore_blob: string | null;
  updated_at?: string;
}

export const worldApi = {
  getLore: async (userId: string, projectId?: number): Promise<WorldLore | null> => {
    return apiRequest<WorldLore>(`${API_BASE}/lore/${userId}`, {
      method: 'GET',
      label: 'Get World Lore',
      headers: projectId ? { 'X-Project-Id': projectId.toString() } : {}
    });
  },

  updateLore: async (userId: string, update: Partial<WorldLore>, projectId?: number): Promise<WorldLore> => {
    return apiRequest<WorldLore>(`${API_BASE}/lore/${userId}`, {
      method: 'POST',
      label: 'Update World Lore',
      body: JSON.stringify(update),
      headers: projectId ? { 'X-Project-Id': projectId.toString() } : {}
    });
  },

  getHistory: async (userId: string, limit: number = 10): Promise<WorldLore[]> => {
    return apiRequest<WorldLore[]>(`${API_BASE}/history/${userId}?limit=${limit}`, {
      method: 'GET',
      label: 'Get World Lore History'
    });
  }
};



